Sources/SocksCore/UDPSocket.swift

```diff
@@ -0,0 +1,83 @@
+//
+//  UDPSocket.swift
+//  Socks
+//
+//  Created by Honza Dvorsky on 6/1/16.
+//
+//
+
+#if os(Linux)
+    import Glibc
+    private let socket_recvfrom = Glibc.recvfrom
+    private let socket_sendto = Glibc.sendto
+#else
+    import Darwin
+    private let socket_recvfrom = Darwin.recvfrom
+    private let socket_sendto = Darwin.sendto
+#endif
+
+public class UDPSocket: InternetSocket {
+    
+    public let descriptor: Descriptor
+    public let config: SocketConfig
+    public let address: ResolvedInternetAddress
+    
+    public required init(descriptor: Descriptor?, config: SocketConfig, address: ResolvedInternetAddress) throws {
+        
+        if let descriptor = descriptor {
+            self.descriptor = descriptor
+        } else {
+            self.descriptor = try UDPSocket.createNewSocket(config: config)
+        }
+        self.config = config
+        self.address = address
+    }
+    
+    public convenience init(address: InternetAddress) throws {
+        var config: SocketConfig = .UDP(addressFamily: address.addressFamily)
+        let resolved = try address.resolve(with: config)
+        config = try config.adjusted(for: resolved)
+        try self.init(descriptor: nil, config: config, address: resolved)
+    }
+    
+    public func recvfrom(maxBytes: Int = BufferCapacity) throws -> (data: [UInt8], sender: ResolvedInternetAddress) {
+        let data = Bytes(capacity: maxBytes)
+        let flags: Int32 = 0 //FIXME: allow setting flags with a Swift enum
+        
+        var length = socklen_t(sizeof(sockaddr_storage))
+        let addr = UnsafeMutablePointer<sockaddr>.init(allocatingCapacity: 1)
+        
+        let receivedBytes = socket_recvfrom(
```

Comment:
Here, the client could potentially be stuck forever when a UDP packet from the server gets lost or corrupted on its way. Adding a timeout or "that's something for the application layer to handle"?